File: signalr-easy-chat-client/src/app/services/signalr.service.ts
import { Injectable } from '@angular/core';
import { environment } from 'src/environments/environment';
import * as signalR from '@microsoft/signalr';
import { BehaviorSubject, catchError, Observable, Subject } from 'rxjs';
import { HttpClient } from '@angular/common/http';

interface Message {
  messageOf: MessageOfEnum,
  messageText: string,
  senderUserName: string
}

export enum MessageOfEnum {
  Sender,
  Receiver
}

@Injectable({
  providedIn: 'root'
})
export class SignalRService {
  public connection: signalR.HubConnection | undefined;
  private messages: BehaviorSubject<Message[]> = new BehaviorSubject<Message[]>([]);
  public messages$: Observable<Message[]> = this.messages.asObservable();

  constructor(
    private http: HttpClient
  ) { }

  async connect(userName: string) {
    if (!userName) {
      return;
    }

    this.connection = new signalR.HubConnectionBuilder()
      .configureLogging(signalR.LogLevel.Information)
      .withUrl(environment.signalRHost + `?userName=${userName}`)
      .withAutomaticReconnect()
      .build();

    this.connection.start()
      .then(function () {
        console.log('SignalR Connected!');
        window.alert("Connect success!")
      }).catch(function (err: any) {
        console.error(err.toString());
        window.alert("Connect Error!")
      });
  }

  async disconnect() {
    await this.connection?.stop();
  }


  async sendMessage(receiverUserName: string, message: string, senderUserName: string) {
    await this.connection?.invoke("SendMessage", receiverUserName, message, senderUserName).catch(ex => console.error(ex));
    const _message: Message = {
      messageOf: MessageOfEnum.Sender,
      messageText: message,
      senderUserName: senderUserName
    }
    this.messages.next([...this.messages.value, _message])
  }

  async sendMessageByCallAPI(receiverUserName: string, message: string, senderUserName: string) {
    const param = {
      receiverUserName: receiverUserName,
      message: message,
      senderUserName: senderUserName
    }
    this.http.post<any>(environment.apiHost + "/SendMessage", param)
      .pipe(catchError(async (ex) => console.error(ex)))
      .subscribe(data => {
        const _message: Message = {
          messageOf: MessageOfEnum.Sender,
          messageText: message,
          senderUserName: senderUserName
        }
        this.messages.next([...this.messages.value, _message])
      });
  }

  addMessageListener() {
    this.connection?.on("ReceiveMessage", (senderUserName: string, message: string) => {
      console.log(message)
      const _message: Message = {
        messageOf: MessageOfEnum.Receiver,
        messageText: message,
        senderUserName: senderUserName
      }
      this.messages.next([...this.messages.value, _message]
      )
    });
  }
}

File: signalr_easy_chat_flutter/lib/src/services/signalr_service.dart
import 'package:signalr_easy_chat_flutter/src/models/message_model.dart';
import 'package:signalr_easy_chat_flutter/src/utils/constant.dart';
import 'package:signalr_netcore/signalr_client.dart';
import 'package:rxdart/rxdart.dart';

class SignalRService {
  late HubConnection? hubConnection;
  late String _userName;
  final BehaviorSubject<List<MessageModel>> messages =
      BehaviorSubject<List<MessageModel>>.seeded([]);
  List<MessageModel> _messages = [];

  Future<void> connect(String userName) async {
    _userName = userName;
    hubConnection = HubConnectionBuilder()
        .withUrl(Constant.hubConnectionUrl + "?userName=$_userName")
        .build();
    hubConnection?.onclose(({Exception? error}) => print("Connection Closed"));

    return hubConnection!.start()?.catchError(
        (error) => {print("Connection error $error"), throw error});
  }

  Future<void> disconnect() async {
    _messages = [];
    messages.add(_messages);

    if (hubConnection != null) {
      await hubConnection?.stop();
      hubConnection = null;
    }
  }

  Future<void> sendMessage(String receiverUserName, String message) async {
    await hubConnection?.invoke("SendMessage",
        args: <String>[receiverUserName, message, _userName]);
    _messages.add(MessageModel(
        messageOf: MessageOfEnum.Sender,
        messageText: message,
        senderUserName: _userName));
    messages.add(_messages);
  }

  void addMessageListener() async {
    hubConnection?.on("ReceiveMessage", _handleReceiveMessage);
  }

  void _handleReceiveMessage(List<Object>? parameters) {
    print(parameters?.toString());
    _messages.add(MessageModel(
        messageOf: MessageOfEnum.Receiver,
        messageText: parameters != null ? parameters[1].toString() : '',
        senderUserName: parameters != null ? parameters[0].toString() : ''));
    messages.add(_messages);
  }
}

SignalRService signalRService = SignalRService();
